### PR TITLE
Fuse Binary and Variadic Elementwise Op without broadcast

### PR DESCRIPTION
**Pull request by chentong319**
This PR tries to fuse binary and variadic op, on top of previous support for unary Ops.
Op with more than one inputs may have extra issues, compared to unary ops:
1. The inputs, other than the one coming from the fusion chain, may be defined after the root Op for fusion. Currently, there is no data dependence analysis and code motion for fusion. The other inputs are required to be block argument, or constant (after canonicalization) so that they are guaranteed to locate before the root Op for correct SSA.
2. Broadcasting may occur for multiple inputs Op.  Consequently loop iteration space and the load index may be different. I think I should use the broadcast shape helper and symbolic shape analysis to handle this . But this PR only handles Op with static shape without broadcast.
3. To load the other inputs, getRemappedValue has to be used to hook with the transformed inputs explicitly. When OpConversionPattern is used, the transformed inputs are given as parameter for matchAndRewrite.

@AlexandreEichenberger and @tungld, when you review the PR, please give suggestion on what utility function to use for general cases. I left the code with current BroadcastShapeHelper in `#if 0`. There is an issue for the getRemappedValue for the fused input, which I do not need to use but the ShapeHelper seems need that.
Fusion may use those Helper function in a way different from our previous lowering. Your knowledge with ShapeHelper may be helpful. For example, I need to compare whether two Ops outputs are the same size, in the sense of our hasNoBroadcast()meaning. 1x5 and 5 is the same. And also one is the producer of the other's one input. I do not whether this condition makes any difference. Which utility function to use?

### CODE REVIEW

**Comment by chentong319** [comment]
I copy-paste a test case, which is included in lit test, for reviewer to read easily. 
Input:
```
// fusible binary with constant input
func.func @fuse_element_14(%arg0: tensor<5xf32>) -> tensor<*xf32> {
  %cst = onnx.Constant dense<[1.0, 2.0, 3.0, 4.0, 5.0]> : tensor<5xf32>
  %0 = "onnx.Sqrt"(%arg0) : (tensor<5xf32>) -> tensor<?xf32>
  %1 = "onnx.Add"(%0, %cst) : (tensor<?xf32>, tensor<5xf32>) -> tensor<*xf32>
  return %1 : tensor<*xf32>
}
```
Output with ` onnx-mlir-opt --shape-inference --convert-onnx-to-krnl --canonicalize`
```
  func.func @fuse_element_14(%arg0: memref<5xf32>) -> memref<5xf32> {
    %0 = "krnl.global"() {name = "constant_0", shape = [5], value = dense<[1.000000e+00, 2.000000e+00, 3.000000e+00, 4.000000e+00, 5.000000e+00]> : tensor<5xf32>} : () -> memref<5xf32>
    %alloc = memref.alloc() {alignment = 16 : i64} : memref<5xf32>
    %1 = krnl.define_loops 1
    krnl.iterate(%1) with (%1 -> %arg1 = 0 to 5){
      %2 = krnl.get_induction_var_value(%1) : (!krnl.loop) -> index
      %3 = krnl.load %arg0[%2] : memref<5xf32>
      %4 = math.sqrt %3 : f32
      %5 = krnl.load %0[%2] : memref<5xf32>
      %6 = arith.addf %4, %5 : f32
      krnl.store %6, %alloc[%2] : memref<5xf32>
    }
    return %alloc : memref<5xf32>
  }
```

**Review by tungld — COMMENTED** [review]
Reading the source code I see many different terms that potentially refer to the same thing: `root op`, `op`, `start op`, `user op`. Could you revise them and make them consistent? Since you use producer-consumer relationship, I guess those terms refer to `consumer`.

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:1542]: I started with the declaration of class in a separate .hpp file and the `mlir::` is needed. Since the Elementwise templates are not in .hpp file, I have to keep these code in Elementwise.cpp now. However, these code eventually will be moved into a separate .hpp file.

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp]: Changed

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp]: I change all the variables to useOp and defOp  since MLIR uses define and use and it's not too long. I switched the parameter order because the useOp is usually the focus of the function, kind of the current op.

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp]: root op and rootOp_ are used.

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
Comment (tungld) [inline src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:1696]: `if (!isa<BlockArgument>(useOp->getOperand(i))` should be called before `Operation *input = useOp->getOperand(i).getDefiningOp();`. Otherwise, `input` can be null. For example,

```c
   Value inputVal = useOp->getOperand(i);

    // Only input from block argument and constant is allowed.
    if (!isa<BlockArgument>(inputVal)) {
        Operation *inputOp = inputVal.getDefiningOp();
        if (inputOp == defOp)
            continue;
        if(!isa<ONNXConstantOp>(input))
            return false;
    }
```

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
Comment (tungld) [inline src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:1537]: Could it be more concrete here? It appears that not all elementwise ops are supported, e.g, two elementwise binary ops, like `Add -> Sub`. I would expect to see a list like:
```
- Pattern 1: unary op -> unary op
- Pattern 2: binary op -> unary op, .... requirements
- Pattern 3: unary op -> binary op, .... requirements
- other patterns ....
```

Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:1537]: Add and Sub are in variadic Op list. I just followed the ONNXElementwiseUnaryOpLowering<T>, ONNXElementwiseBinaryOpLowering<T>, and ONNXElementwiseVariadicOpLowering<T>.
They are all supported, except ClipOp. ClipOp is in ONNXElementwiseUnaryOpLowering, but it has three input and does not support broadcast. In fact, such op could be fused too.

Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:1537]: The condition for unary Op and binary Op are the same. But the unary is simpler  because it has only the input to be  fused, no extra loads for other input.

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp]: Changed.
Yes, useOp->getOperand(i) could be null. But the test case works. Does it mean that null object is fine?

**Review by tungld — APPROVED** [review]
LGTM, and please wait for @AlexandreEichenberger's review before merging it.

**Review by AlexandreEichenberger — COMMENTED** [review]
Please address Tung's suggestions and mine. Looks generally good for its intended purpose.

We should talk on when to start thinking about integration with SIMD.

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:1782]: Function parameter, or Loop body parameter, are examples of BlockArgument. They are defined at the top of the current block. I will add a check that the define and use Op should be in the same block. Fusion across block may not only violate the dependence. but also may break the control flow.

Here is an example. `%arg1` is a block argument and is an input for Add op.
```
func.func @fuse_element_10(%arg0: tensor<4x5xf32>, %arg1: tensor<4x5xf32>) -> tensor<*xf32> {
  %0 = "onnx.Sqrt"(%arg0) : (tensor<4x5xf32>) -> tensor<4x5xf32>
  %1 = "onnx.Add"(%0, %arg1) : (tensor<4x5xf32>, tensor<4x5xf32>) -> tensor<*xf32>
  return %1 : tensor<*xf32>
}
```

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:1621]: It's just a coding style to attach `_` for class member variable.  I can move the variable to the bottom class.

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:1720]: For the template, let me explain again.  In fusion we need to check the Op of all type, while the Element-wise lowering handles only the element-wise ops. Element-wise lowering start from element-wise Op and therefore it is valid to reference ScalarOp.  If we define a krnlLoweringOpInterface for all ONNX Op, element-wise property will be part of it, and then we can check the Op with defined methods for the KrnlLoweringOpInterface. This template is not needed.

Agree on the two-phase plan.

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:1796]: @chentong319 did you reformulate the paragraph?

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:1812]: Thanks for the explanation, it feel that we should use this in more places.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM after you edit or confirm that you edited the paragraph requested by @tungld

**Pull request merged**
